src/core/response_cache.cc

```diff
@@ -0,0 +1,284 @@
+            Status::Code::INTERNAL, "key not found in cache"
+        );
+    }
+    // Update this key to front of LRU list
+    UpdateLRU(iter);
+    // Populate passed-in "ptr" from cache entry
+    auto entry = iter->second;
+    // Build InferenceResponse from CacheEntry
+    auto status = BuildInferenceResponse(entry, ptr);
+    if (!status.IsOk()) {
+        return status;
+    }
+
+    return Status::Success;
+}
+
+Status RequestResponseCache::Insert(const uint64_t key, const InferenceResponse& response) {
+    // Exit early if key already exists in cache
+    auto iter = cache_.find(key);
+    if (iter != cache_.end()) {
+        return Status(
+            Status::Code::INTERNAL, "key already exists in cache"
+        );
+    }
+    // Construct cache entry from response
+    auto entry = CacheEntry();
+    auto status = BuildCacheEntry(entry, response);
+    if (!status.IsOk()) {
+        return status;
+    }
+
+    // If cache doesn't have room for new entry, evict until enough size is available
+    while (entry.size > available_size_) {
+        auto status = Evict();
+        // If evict fails for some reason, exit with its failure status
+        if (!status.IsOk()) {
+            return status;
+        }
+    }
+    // Insert entry into cache
+    auto cache_pair = cache_.insert({key, entry});
+    bool ok = cache_pair.second;
+    // Exit early if cache insertion failed
+    if (!ok) {
+        return Status(
+            Status::Code::INTERNAL, "Cache insertion failed"
+        );
+    }
+    // Update LRU with new cache entry
+    auto cache_iter = cache_pair.first;
+    UpdateLRU(cache_iter);
+    // Update available cache size
+    available_size_ -= entry.size;
+    // Error checking cache size management
+    if (available_size_ > total_size_) {
+        return Status(
+            Status::Code::INTERNAL,
+            "Available size exceeded total size: this indicates a bug in the code"
+        );
+    }
+
+    return Status::Success;
+}
+
+Status RequestResponseCache::BuildCacheEntry(CacheEntry& entry, const InferenceResponse& response) {
+    // Build cache entry data from response outputs
+    //for (const InferenceResponse::Output& response_output : response.Outputs()) {
+    for (const auto& response_output : response.Outputs()) {
+        auto cache_output = Output();
+
+        // Fetch output buffer details
+        const void* response_buffer = nullptr;
+        size_t response_byte_size = 0;
+        /*
+        TRITONSERVER_MemoryType response_memory_type;
+        int64_t response_memory_type_id = 0;
+        void* userp = nullptr;
+        // TODO: How to differently handle different memory types?
+        //       GPU vs. CPU memory, etc.
+        // TODO: Linker error on calling DataBuffer() here
+        Status status = response_output.DataBuffer(
+            &response_buffer, &response_byte_size, &response_memory_type,
+            &response_memory_type_id, &userp);
+        */
```

Comment:
Simple Hello World unit test in `src/test/response_cache_test.cc` with `#include "src/core/response_cache.h"` able to compile with calls to `InferenceResponse::Output::DataBuffer()` commented out:
```
root@64a5c57847e5:/mnt/jira/2624-cache/server/build/build/test-util/install/bin# ./response_cache_test 
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from RequestResponseCacheTest
[ RUN      ] RequestResponseCacheTest.TestHelloWorld
Hello, World!
[       OK ] RequestResponseCacheTest.TestHelloWorld (0 ms)
[----------] 1 test from RequestResponseCacheTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.
```

Reply:
This is where you will provide mock implementations for the classes that are not under test. The [repo agent unit test](https://github.com/triton-inference-server/server/blob/main/src/test/triton_repo_agent_test.cc#L190-L249) is one example where it provides mock implementation for the [SharedLibrary class](https://github.com/triton-inference-server/server/blob/main/src/core/shared_library.h), one benefit is that the mock implementation doesn't involve file system and actual shared library files, so it is easy to provide various function handlers as if they are from shared library in the real use case.

The request / response class header exposes some implementation details so it can be harder than mocking SharedLibrary class, but I think you can skip some of them as long as they are not used in your unit test.